tiledb/api/src/range.rs

```diff
@@ -0,0 +1,841 @@
+            Self::Int16(_, _) => check_datatype_inner!(i16, $datatype),
+            Self::Int32(_, _) => check_datatype_inner!(i32, $datatype),
+            Self::Int64(_, _) => check_datatype_inner!(i64, $datatype),
+            Self::Float32(_, _) => check_datatype_inner!(f32, $datatype),
+            Self::Float64(_, _) => check_datatype_inner!(f64, $datatype),
+        }
+    };
+}
+
+#[derive(Clone, Deserialize, Serialize, PartialEq)]
+pub enum SingleValueRange {
+    UInt8(u8, u8),
+    UInt16(u16, u16),
+    UInt32(u32, u32),
+    UInt64(u64, u64),
+    Int8(i8, i8),
+    Int16(i16, i16),
+    Int32(i32, i32),
+    Int64(i64, i64),
+    Float32(f32, f32),
+    Float64(f64, f64),
+}
+
+impl SingleValueRange {
+    pub fn check_datatype(&self, datatype: Datatype) -> TileDBResult<()> {
+        check_datatype!(self, datatype);
+        Ok(())
+    }
+}
+
+macro_rules! single_value_range_from {
+    ($($V:ident : $U:ty),+) => {
+        $(
+            impl From<&[$U; 2]> for SingleValueRange {
+                fn from(value: &[$U; 2]) -> SingleValueRange {
+                    SingleValueRange::$V(value[0], value[1])
+                }
+            }
+        )+
+    }
+}
+
+single_value_range_from!(UInt8: u8, UInt16: u16, UInt32: u32, UInt64: u64);
+single_value_range_from!(Int8: i8, Int16: i16, Int32: i32, Int64: i64);
+single_value_range_from!(Float32: f32, Float64: f64);
+
+#[macro_export]
+macro_rules! single_value_range_go {
+    ($expr:expr, $DT:ident, $start:pat, $end:pat, $then:expr) => {
+        match $expr {
+            SingleValueRange::UInt8($start, $end) => {
+                type $DT = u8;
+                $then
+            }
+            SingleValueRange::UInt16($start, $end) => {
+                type $DT = u16;
+                $then
+            }
+            SingleValueRange::UInt32($start, $end) => {
+                type $DT = u32;
+                $then
+            }
+            SingleValueRange::UInt64($start, $end) => {
+                type $DT = u64;
+                $then
+            }
+            SingleValueRange::Int8($start, $end) => {
+                type $DT = i8;
+                $then
+            }
+            SingleValueRange::Int16($start, $end) => {
+                type $DT = i16;
+                $then
+            }
+            SingleValueRange::Int32($start, $end) => {
+                type $DT = i32;
+                $then
+            }
+            SingleValueRange::Int64($start, $end) => {
+                type $DT = i64;
+                $then
+            }
+            SingleValueRange::Float32($start, $end) => {
+                type $DT = f32;
+                $then
+            }
+            SingleValueRange::Float64($start, $end) => {
+                type $DT = f64;
+                $then
+            }
+        }
+    };
+}
+
+#[derive(Clone, Deserialize, Serialize, PartialEq)]
+pub enum MultiValueRange {
+    UInt8(Box<[u8]>, Box<[u8]>),
```

Comment:
I recall you prefer `Box` over `Vec` to avoid the excess capacity.  But a convenient way to create this sort of thing is by creating a `Vec` and then calling `into_boxed_slice`.

I was curious if an additional `truncate` would be required, but the reference doc for `into_boxed_slice` says it already does that - great!

Reply:
`Box` over `Vec` isn't because of excess capacity (though it doesn't hurt), its because if we grab pointers to slices derived from Vec slices and pass them across the ffi boundary, they'll be invalidated as soon as the `Vec` reallocates which scared the dickens out of me. Using boxed slices means that as long as we have a ref to it, we're guaranteed that the underlying pointer is valid.

Reply:
(To be clear I do not object to the use of `Box`)

But we don't really keep pointers to the sliced thing for long, do we?  Doesn't the C API copy stuff out of whatever we gave it?

Reply:
True for the most part, except for the big ones for query buffers. We've also done the thing where we re-set each of those pointers on every call to be every sure as well which is good for defense in depth.